src/coreclr/src/jit/flowgraph.cpp

```diff
@@ -23579,10 +23579,14 @@ Statement* Compiler::fgInlinePrependStatements(InlineInfo* inlineInfo)
 
     CORINFO_METHOD_INFO* InlineeMethodInfo = InlineeCompiler->info.compMethodInfo;
 
-    unsigned lclCnt = InlineeMethodInfo->locals.numArgs;
+    unsigned lclCnt    = InlineeMethodInfo->locals.numArgs;
+    bool     bbInALoop = (block->bbFlags & BBF_BACKWARD_JUMP) != 0;
```

Comment:
We set `BBF_BACKWARD_JUMP` very early, just based on lexical IL. So it may encompass non-loop blocks (eg loop with a branch that leads to a `return`). 

We may also optimize away backwards branches (probably rare, but not impossible).

I wonder how often we end up with extra zeroing from `BBF_BACKWARD_JUMP` being approximate? Any Idea?

Reply:
No, I don't know. I guess I can try to allow this optimization for `BBJ_RETURN` basic blocks in root methods since we can execute those blocks only once even in a loop. That won't account for all cases you are asking about but I don't see an easy way to capture those.

Reply:
To estimate how often this happens, you could perhaps set a new BBF in blocks where we add zero inits. Then later when we run some more robust form of loop detection, see how many non-loop blocks have this BBF.